mm2src/mm2_metrics/src/native.rs

```diff
@@ -0,0 +1,509 @@
+        }
+    }
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct TagMetric {
+    pub tags: Vec<Tag>,
+    pub message: String,
+}
+
+#[derive(PartialEq, PartialOrd)]
+pub(crate) enum PreparedMetric {
+    Unsigned(u64),
+    Float(f64),
+    Histogram(MmHistogram),
+}
+
+pub struct TagObserver;
+
+impl TagObserver {
+    /// This is for collecting and logging `prepare_tag_metrics` to dashboard.
+    ///
+    /// Used with `init_with_dashboard`.
+    pub async fn log_tag_metrics(log_state: LogWeak, recorder: Arc<MmRecorder>, interval: f64) {
+        loop {
+            Timer::sleep(interval).await;
+            let log_state = match LogArc::from_weak(&log_state) {
+                Some(la) => la,
+                _ => {
+                    return;
+                },
+            };
+
+            log!(">>>>>>>>>> DEX metrics <<<<<<<<<");
+
+            Self::prepare_tag_metrics(&recorder)
+                .into_iter()
+                .for_each(|(labels, name_value_map)| {
+                    let tags = labels_to_tags(labels.iter());
+                    let message = name_value_map_to_message(&name_value_map);
+                    log_state.log_deref_tags("", tags, &message);
+                });
+        }
+    }
+
+    /// Prepare tag metrics for logging in `log_tag_metrics`.
+    fn prepare_tag_metrics(recorder: &MmRecorder) -> HashMap<MetricLabels, MetricNameValueMap> {
+        let mut output = HashMap::new();
+
+        for (key, counter) in recorder.registry.get_counter_handles() {
+            let value = counter.get_inner().load(Ordering::Acquire);
+            map_metrics_to_prepare_tag_metric_output(key, PreparedMetric::Unsigned(value), &mut output);
+        }
+
+        for (key, gauge) in recorder.registry.get_gauge_handles() {
+            let value = f64::from_bits(gauge.get_inner().load(Ordering::Acquire));
+            map_metrics_to_prepare_tag_metric_output(key, PreparedMetric::Float(value), &mut output);
+        }
+
+        for (key, histo) in recorder.registry.get_histogram_handles() {
+            if let Some(values) = MmHistogram::new(&histo.get_inner().data()) {
+                map_metrics_to_prepare_tag_metric_output(key, PreparedMetric::Histogram(values), &mut output);
+            }
+        }
+        output
+    }
+}
+
+#[derive(PartialEq, PartialOrd)]
+pub(crate) struct MmHistogram {
+    count: usize,
+    min: f64,
+    max: f64,
+}
+
+impl MmHistogram {
+    /// Create new MmHistogram from `&[f64]`.
+    ///
+    /// Return None if data.len() <= 0.
+    pub(crate) fn new(data: &[f64]) -> Option<MmHistogram> {
+        let count: usize = data.len();
+        if count > 0 {
+            let min: f64 = data.iter().cloned().fold(f64::INFINITY, f64::min);
+            let max: f64 = data.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
+            return Some(MmHistogram { count, min, max });
+        }
+        None
```

Comment:
It's worth to avoid using `f64::INFINITY` and `f64::NEG_INFINItY` as initial values.
It's common to consider the first element as min/max value:
```rust
let mut it = data.iter();
let min = data.next()?;
let max = min;
// continue
```
Please also note that it's not efficient to pass through the array elements twice. It could be done in one pass through the elements.

It was theory. Fortunately, there is a library that implemented the searching of min and max: [Itertools::minmax](https://docs.rs/itertools/0.7.8/itertools/trait.Itertools.html#method.minmax). Please note that it returns [MinMaxResult](https://docs.rs/itertools/0.7.8/itertools/enum.MinMaxResult.html) that implements [into_option](https://docs.rs/itertools/0.7.8/itertools/enum.MinMaxResult.html#impl)